### PR TITLE
feat(edit-page): Allow User to Choose Content Editing for Individual or All Pages  #27551

### LINKED ISSUE #27551: Edit Page V2: Allow User to Choose Content Editing for Individual or All Pages
### Parent
#27546 

### User Story
As a content editor, I want to be informed if the content I am editing on a page will affect other pages, so that I can make an informed decision about whether to edit the content for all pages or only for the current page.

image

```[tasklist]
### Acceptance Criteria
- [x] When the user clicks on the "Edit" button for a content item on a page, a dialog should be displayed.
- [x] Dialog only shows if the edit content is being used in another page
- [x] The functionality should be consistent with the existing "Edit" functionality in the old edit page.
- [x] The feature should be implemented in the new edit page.
- [x] This functionality should work for headless pages as well
```

### Proposed Objective
Static content: "Core Features"

### Proposed Priority
Static content: "Priority 3 - Average"

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.
Static content: "N/A"

### PR DESCRIPTION
**Pull request by rjvelazco**
- Allow users to choose content editing for individual or all pages.

### Video

https://github.com/dotCMS/core/assets/72418962/34a93f6f-4876-4553-8756-04148510b04f

#### If copying the content fails

https://github.com/dotCMS/core/assets/72418962/e22adc75-75eb-4fa7-9d38-30a0ec7c7dab

### CODE REVIEW

File: core-web/libs/sdk/experiments/.eslintrc.json [hunk 1]
```diff
@@ -1,33 +1,33 @@
...
-          {
-            "selector": "class",
-            "format": ["PascalCase"],
-            "leadingUnderscore": "allow",
-            "trailingUnderscore": "allow"
-          }
-        ]
-      }
-    },
+    "extends": ["../../../.eslintrc.json"],
```
Comment (oidacra) [inline core-web/libs/sdk/experiments/.eslintrc.json:2]: Try to ignore those files, Prettier doing disasters.

File: core-web/libs/ui/src/lib/services/dot-copy-content-modal/dot-copy-content-modal.service.ts [hunk 2]
```diff
@@ -47,7 +48,7 @@ export class DotCopyContentModalService {
      * @memberof DotEditContentComponent
      * @returns {Observable<ModelCopyContentResponse>}
      */
-    open(): Observable<ModelCopyContentResponse> {
+    open(onClose?: () => void): Observable<ModelCopyContentResponse> {
```
Comment (oidacra) [inline core-web/libs/ui/src/lib/services/dot-copy-content-modal/dot-copy-content-modal.service.ts:51]: TIL.

File: core-web/libs/sdk/experiments/.eslintrc.json
Comment (zJaaal) [inline core-web/libs/sdk/experiments/.eslintrc.json:1]: you need to fix prettier and make a commit in order to husky to re-format all the files and clean this PR

File: core-web/libs/portlets/edit-ema/portlet/src/lib/shared/models.ts [hunk 3]
```diff
@@ -6,6 +8,8 @@ export interface ActionPayload {
     pageId: string;
     personaTag?: string;
     position?: 'before' | 'after';
+    treeNode?: DotTreeNode;
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/shared/models.ts:11]: I changed this type in my [PR](https://github.com/dotCMS/core/pull/27625), is kind of cleaner now

File: core-web/libs/sdk/react/src/lib/components/Container/Container.tsx [hunk 4]
```diff
@@ -115,10 +115,20 @@ export function Container({ containerRef }: ContainerProps) {
...
             },
+            treeNode: {
+                containerId: container.identifier,
+                contentId: contentlet.identifier,
+                relationType: container.uuid,
+                variantId: container.variantId,
+                personalization: getPersonalization(viewAs.persona),
+                treeOrder,
+                pageId
+            },
```
Comment (zJaaal) [inline core-web/libs/sdk/react/src/lib/components/Container/Container.tsx:126]: we already has this information in the container in LN122, check if we can make use of that.

File: core-web/libs/sdk/react/src/lib/utils/utils.ts [hunk 5]
```diff
@@ -140,10 +141,19 @@ export const getContainersData = (
         ...containers[identifier].container,
         acceptTypes,
         contentlets,
-        pageContainers
+        pageContainers,
+        variantId
     };
 };
```
Comment (zJaaal) [inline core-web/libs/sdk/react/src/lib/utils/utils.ts:148]: Docs

**Review by zJaaal — COMMENTED** [review]
Try to clean the prettier format errors and also I'm not sure how are we going to merge our branches, I'm not sure if we can get this info from the store instead of the client.

[PR](https://github.com/dotCMS/core/pull/27625)

File: core-web/libs/ui/src/lib/services/dot-copy-content-modal/dot-copy-content-modal.service.ts (hunk 2 above)
Comment (rjvelazco) [reply core-web/libs/ui/src/lib/services/dot-copy-content-modal/dot-copy-content-modal.service.ts:51]: It is helpful in some cases

File: core-web/libs/portlets/edit-ema/portlet/src/lib/shared/models.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/shared/models.ts:11]: This should always come. The `?:` was only temporary.

File: core-web/libs/sdk/react/src/lib/utils/utils.ts [hunk 6]
```diff
@@ -96,7 +96,10 @@ export const getContainersData = (
 ) => {
     const { identifier, uuid } = containerRef;
 
-    const { containerStructures } = containers[identifier];
+    const { containerStructures, container } = containers[identifier];
+
+    // Get the variant id
+    const { variantId } = container?.parentPermissionable || {};
```
Comment (manuelrojas) [inline core-web/libs/sdk/react/src/lib/utils/utils.ts:102]: Is variantId a string or number? Can we a the type here?

File: core-web/libs/portlets/edit-ema/portlet/src/lib/utils/index.ts [hunk 7]
```diff
@@ -147,3 +152,37 @@ export function sanitizeURL(url: string): string {
...
+ * @param {ContentletPayload} contentlet
+ * @param {ContainerPayload} container
+ * @return {*}
+ */
+export function getContentletRelationship(
+    contentlet: ContentletPayload,
+    container: ContainerPayload
+) {
+    return {
+        contentId: contentlet?.identifier,
```
Comment (manuelrojas) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/utils/index.ts:169]: why only in the first one we are doing `contentlet?.`

File: core-web/libs/portlets/edit-ema/portlet/src/lib/components/dot-ema-dialog/dot-ema-dialog.component.ts [hunk 8]
```diff
@@ -187,4 +201,31 @@ export class DotEmaDialogComponent {
...
+     * @return {*}
+     * @memberof DotEmaDialogComponent
+     */
+    private askToCopy({ treeNode, contentlet }) {
+        return this.dotCopyContentModalService
+            .open(() => this.loading.emit(true))
+            .pipe(
+                switchMap(({ shouldCopy }) => {
+                    return shouldCopy
+                        ? this.dotCopyContentService.copyInPage(treeNode)
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/components/dot-ema-dialog/dot-ema-dialog.component.ts:219]: I was working on this! Good catch, tho!

File: core-web/libs/sdk/react/src/lib/components/Container/Container.tsx (hunk 4 above)
Comment (rjvelazco) [reply core-web/libs/sdk/react/src/lib/components/Container/Container.tsx]: Done!

File: core-web/libs/portlets/edit-ema/portlet/src/lib/components/dot-ema-dialog/dot-ema-dialog.component.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/components/dot-ema-dialog/dot-ema-dialog.component.ts]: @manuelrojas 

https://github.com/dotCMS/core/assets/72418962/e22adc75-75eb-4fa7-9d38-30a0ec7c7dab

File: core-web/libs/sdk/react/src/lib/utils/utils.ts (hunk 6 above)
Comment (rjvelazco) [reply core-web/libs/sdk/react/src/lib/utils/utils.ts:102]: The parentPermissionable is a `Record<string, string>`

Here: https://github.com/dotCMS/core/blob/48547c9d4b4b10cdbd80c98d92554dbb4bdec0b6/core-web/libs/sdk/react/src/lib/components/PageProvider/PageProvider.tsx#L17

**Review by manuelrojas — APPROVED** [review]

File: core-web/libs/sdk/experiments/.eslintrc.json
Comment (rjvelazco) [reply core-web/libs/sdk/experiments/.eslintrc.json:1]: Done!

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 9]
```diff
@@ -625,6 +643,17 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
 
+    /**
+     * Update editor state to loading
+     *
+     * @protected
+     * @param {boolean} isLoading
+     * @memberof EditEmaEditorComponent
+     */
+    protected handleLoading(isLoading: boolean) {
+        this.store.updateEditorState(isLoading ? EDITOR_STATE.LOADING : EDITOR_STATE.LOADED);
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:648]: I removed this and followed another approach

**Review by zJaaal — APPROVED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 10]
```diff
@@ -131,24 +135,32 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
+            pageId,
             pageContainers: containers,
             personaTag,
             container: {
-                ...clientData.container,
+                ...container,
                 contentletsId
-            }
+            },
+            treeNode
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:163]: Should we normalize this data? there are some duplications here.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts [hunk 11]
```diff
@@ -641,7 +645,8 @@ describe('EditEmaStore', () => {
                         uuid: '123',
                         acceptTypes: 'test',
                         maxContentlets: 1,
-                        contentletsId: ['existing-contentlet-123', 'form-identifier-123']
+                        contentletsId: ['existing-contentlet-123', 'form-identifier-123'],
+                        variantId: '1'
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts:649]: The `/api/v1/page/copyContent` endpoint requires the `VariantId` to be included in the request body.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/utils/index.ts [hunk 12]
```diff
@@ -147,3 +152,37 @@ export function sanitizeURL(url: string): string {
...
+        containerId: container.identifier,
+        relationType: container.uuid,
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/utils/index.ts:171]: Backend thing. The last time I talked with @jdotcms, he advised me to use the `uuid` as the `relationType` in the copy content request.

File: core-web/libs/sdk/experiments/.eslintrc.json
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/sdk/experiments/.eslintrc.json:1]: Don!

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts (hunk 10 above)
Comment (rjvelazco) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts]: Done!

**Review by fmontes — APPROVED** [review]

**Review by oidacra — APPROVED** [review]

**Pull request merged**